summarize/src/analysis.rs

```diff
@@ -20,141 +74,76 @@ pub fn perform_analysis(data: ProfilingData) -> Results {
-                {
-                    if let Some(prev_event) = thread_stack.last() {
-                        //count the time run so far for this event
-                        let duration = event
-                            .timestamp
-                            .duration_since(prev_event.timestamp)
-                            .unwrap_or(Duration::from_nanos(0));
-
-                        record_event_data(&prev_event.label, &|data| {
-                            data.self_time += duration;
-                        });
-
-                        //record the total time
-                        total_time += duration;
-                    }
-
-                    thread_stack.push(event);
-                } else if &event.event_kind[..] == QUERY_BLOCKED_EVENT_KIND
-                    || &event.event_kind[..] == INCREMENTAL_LOAD_RESULT_EVENT_KIND
-                {
-                    thread_stack.push(event);
-                }
-            }
-            TimestampKind::Instant => {
-                if &event.event_kind[..] == QUERY_CACHE_HIT_EVENT_KIND {
-                    record_event_data(&event.label, &|data| {
+    for current_event in data.iter().rev() {
+        match current_event.timestamp {
+            Timestamp::Instant(_) => {
+                if &current_event.event_kind[..] == QUERY_CACHE_HIT_EVENT_KIND {
+                    record_event_data(&current_event.label, &|data| {
                         data.number_of_cache_hits += 1;
                         data.invocation_count += 1;
                     });
                 }
             }
-            TimestampKind::End => {
-                let thread_stack = threads.get_mut(&event.thread_id).unwrap();
-                let start_event = thread_stack.pop().unwrap();
-
-                assert_eq!(start_event.label, event.label);
-                assert_eq!(start_event.event_kind, event.event_kind);
-                assert_eq!(start_event.timestamp_kind, TimestampKind::Start);
-
-                //track the time for this event
-                let duration = event
-                    .timestamp
-                    .duration_since(start_event.timestamp)
-                    .unwrap_or(Duration::from_nanos(0));
-
-                if &event.event_kind[..] == QUERY_EVENT_KIND
-                    || &event.event_kind[..] == GENERIC_ACTIVITY_EVENT_KIND
-                {
-                    record_event_data(&event.label, &|data| {
-                        data.self_time += duration;
-                        data.number_of_cache_misses += 1;
-                        data.invocation_count += 1;
+            Timestamp::Interval { .. } => {
+                // This is an interval event
+                let thread_stack = threads.entry(current_event.thread_id).or_default();
+
+                // Pop all events from the stack that are not parents of the
+                // current event.
+                while let Some(current_top) = thread_stack.last().cloned() {
```

Comment:
Why would we have events on the stack that aren't parents of the current event?

Reply:
Since we don't explicit "end" events anymore, this is now the only mechanism for popping things. Consider the following situation:
```
     <--e2-->   <--e3-->
 <-----------e1----------->
```
In the event stream this shows up as something like:
```
[
	{ label=e2, start= 5, end=10 },
	{ label=e3, start=15, end=20 },
	{ label=e1, start= 0, end=25 },
]
```
because events are emitted in the order of the end timestamps.  So, as we walk backwards, we

1. encounter `e1`, push it onto the stack, then
2. encounter `e3`, the stack contains `e1`, but that is fine since the time-interval of `e1` includes the time interval of `e3`. `e3` goes onto the stack and then we
3. encounter `e2`. The stack is `[e1, e3]`, but now `e3` needs to be popped because we are past its range, so we pop `e3` and push `e2`.

Why is popping done in a `while` loop? consider the following

```
                 <-e4->
     <--e2-->   <--e3-->
 <-----------e1----------->
```

This looks as follows in the stream:
```
[
	{ label=e2, start= 5, end=10 },
	{ label=e4, start=17, end=19 },
	{ label=e3, start=15, end=20 },
	{ label=e1, start= 0, end=25 },
]
```
In this case when we encounter `e2`, the stack is `[e1, e3, e4]`, and both `e4` and `e3` need to be popped in the same step.

Reply:
Brilliant, thanks!

Reply:
I'll see if I can add the above comment to the docs somewhere.